src/promise.c

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "crt.h"
+
+#include <aws/common/condition_variable.h>
+#include <aws/common/mutex.h>
+
+struct _aws_crt_promise {
+    struct aws_mutex mutex;
+    struct aws_condition_variable cv;
+    bool complete;
+    int error_code;
+    void *value;
+    /* destructor for value, will be invoked if the value is not taken */
+    void (*dtor)(void *);
+};
+
+aws_crt_promise *aws_crt_promise_new(void) {
+    aws_crt_promise *promise = aws_crt_mem_acquire(sizeof(aws_crt_promise));
```

Comment:
We don't have an aws_crt_mem_calloc?  What kind of barbaric api requires memset?!

Reply:
I had this thought as I was writing it. Easy fix.